### PR TITLE
578 ddf all zero in predict

### LINKED ISSUE #578: DDF all zero in predict
## Subject of the issue

In the latest main branch, when using demos/paired_mrus_prostate/demo_predict.py. The generated DDF are all zero.

The issue comes from ba27961deb0a3d1721747e03ae5a25ef49d988ea, #556.

## Steps to reproduce

- Train 1 epoch
- Copy the checkpoint under dataset/pretrained
- Run predict
- Check saved DDF png files

## Expected behaviour

Non-uniform DDF.

## Actual behaviour

All zero DDF.

## TODO

- Fix the problem by using the same loading mechanism as train. But be careful, as we changed the save/load system, the saved ckpts in demos were old versions.
- Add a test, verifying the DDF will not be all zero.
- Test in this demo manually.

### PR DESCRIPTION
**Pull request by kate-sann5100**
# Description

Please include a summary of the change and which issue is fixed. Please also include
relevant motivation and context. List any dependencies that are required for this
change.

Fixes #578 

## Type of change

What types of changes does your code introduce to DeepReg?

_Please check the boxes that apply after submitting the pull request._

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Documentation Update (fix or improvement on the documentation)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Other (if none of the other choices apply)

## Checklist

_Please check the boxes that apply after submitting the pull request._

_If you're unsure about any of them, don't hesitate to ask. We're here to help! This is
simply a reminder of what we are going to look for before merging your code._

- [x] I have
      [installed pre-commit](https://deepreg.readthedocs.io/en/latest/contributing/setup.html)
      using `pre-commit install` and formatted all changed files. If you are not
      certain, run `pre-commit run --all-files`.
- [x] My commits' message styles matches
      [our requested structure](https://deepreg.readthedocs.io/en/latest/contributing/commit.html),
      e.g. `Issue #<issue number>: detailed message`.
- [ ] I have updated the
      [change log file](https://github.com/DeepRegNet/DeepReg/blob/main/CHANGELOG.md)
      regarding my changes.
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by YipengHu** [comment]
@kate-sann5100 is this one ready for review now? i have two branches open with different format checkpoints that i can merge to check quickly.

File: deepreg/callback.py [hunk 1]
```diff
@@ -1,38 +1,39 @@
...
-    return [tensorboard_callback, checkpoint_manager_callback]
+    if ckpt_path:
+        initial_epoch = int(ckpt_path.split("-")[-1])
+        checkpoint_manager_callback.restore(ckpt_path)
+    else:
+        initial_epoch = 0
+    return checkpoint_manager_callback, initial_epoch
 
 
 def restore_model(callbacks, ckpt_path):
```
Comment (mathpluscode) [inline deepreg/callback.py:33]: remove this func as i think its no longer used

File: deepreg/predict.py [hunk 2]
```diff
@@ -316,7 +317,17 @@ def predict(
 
     # load weights
     # https://stackoverflow.com/questions/58289342/tf2-0-translation-model-error-when-restoring-the-saved-model-unresolved-objec
-    model.load_weights(ckpt_path).expect_partial()
+    # model.load_weights(ckpt_path).expect_partial()
```
Comment (mathpluscode) [inline deepreg/predict.py:320]: remove commented code

File: demos/paired_mrus_prostate/demo_predict.py [hunk 3]
```diff
@@ -41,7 +41,8 @@
 
 log_root = f"demos/{name}"
 log_dir = "logs_predict/" + datetime.now().strftime("%Y%m%d-%H%M%S")
-ckpt_path = f"{log_root}/dataset/pretrained/weights-epoch500.ckpt"
+# ckpt_path = f"{log_root}/dataset/pretrained/weights-epoch500.ckpt"
+ckpt_path = f"{log_root}/dataset/pretrained/ckpt-5000"
```
Comment (mathpluscode) [inline demos/paired_mrus_prostate/demo_predict.py:45]: revert changes

File: test/unit/test_callback.py [hunk 4]
```diff
@@ -38,35 +38,32 @@ def toy_dataset():
         old_model = Net()
         old_optimizer = tf.keras.optimizers.Adam(0.1)
     old_model.compile(optimizer=old_optimizer, loss=tf.keras.losses.MSE)
-    old_callbacks = build_callbacks(
+    old_callback, _ = build_checkpoint_manager_callback(
```
Comment (mathpluscode) [inline test/unit/test_callback.py:41]: use meaningful name, like ckpt_manager_callback

File: demos/paired_mrus_prostate/demo_predict.py (hunk 3 above)
Comment (kate-sann5100) [reply demos/paired_mrus_prostate/demo_predict.py]: done

File: deepreg/predict.py (hunk 2 above)
Comment (kate-sann5100) [reply deepreg/predict.py]: done

File: deepreg/callback.py (hunk 1 above)
Comment (kate-sann5100) [reply deepreg/callback.py]: done

**Comment by kate-sann5100** [comment]
> @kate-sann5100 is this one ready for review now? i have two branches open with different format checkpoints that i can merge to check quickly.

This is ready for review. Please let me know the checking result, Thanks.

**Comment by mathpluscode** [comment]
DDF from demo saved ckpts

image

DDF from @YipengHu ckpts

image

So I confirm the bug is fixed in this branch.

**Review by mathpluscode — APPROVED** [review]

**Review by YipengHu — APPROVED** [review]
tested on #571 .
